### PR TITLE
Performance test for querying all aggregators.

### PR DESCRIPTION
**Pull request by jcdyer**
@symbolist, @bradenmacdonald

OC-4952

tl;dr: On a course with 10,000 users, it takes 10 minutes to query all course-level aggregators.

Larger page sizes shrink the total query time (more pages to fetch, fewer results per page).

page_size=100 (100 pages) -> 1400s
page_size=1000 (10 pages) -> 590s
page_size= 5000 (2 pages) -> 450s

Number of blocks in the course don't seem to affect the query time.

Checkout the current branch. Run the following commands (in a solutions devstack):

    $ pip uninstall openedx-completion-aggregator
    $ pip install git+https://github.com/open-craft/openedx-completion-aggregator@master#egg=openedx-completion-aggregator
    $paver test_system -s lms -t openedx/tests/aggregator_load -e '--logging-level=WARNING --nocapture'

Tweak values for number of users and page_size in `openedx/tests/aggregator_load/__init__.py`

Not super happy about this running time. A quick view of the aggregator code suggests that we can probably improve this in the way we're using AggregatorAdapter. It seems to force us to iterate over the queryset for each enrollment, rather than associating each Aggregator with the appropriate enrollment on a single pass over the queryset.

One limitations of this work is that there are no Aggregators that aren't part of the course being queried. Due to indexing on the aggregator table, this shouldn't be a problem, but it would be nice to test it.

Also, queries are run sequentially. They could be parallelized, but as always, using more server resources for a single view takes more away from other users, so that might not be wise.

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @jcdyer! I've created [OSPR-2525](https://openedx.atlassian.net/browse/OSPR-2525) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

**Comment by jcdyer** [comment]
Depending on page_size and user_count parameters (and this is page_size=5000, user_count=10000) results look like: 

```
CREATING USERS 2018-07-16 21:43:07.470893+00:00
CREATING ENROLLMENTS AND AGGREGATORS 2018-07-16 21:44:47.364769+00:00
START 2018-07-16 21:44:53.170979+00:00
[]
   {'count': 10000, 'previous': None, 'num_pages': 2, 'next': 'http://testserver/api/completion-aggregator/v1/course/org.0/course_0/Run_0/?page=2&page_size=5000'} 2018-07-16 21:48:30.341737+00:00
   {'count': 10000, 'previous': 'http://testserver/api/completion-aggregator/v1/course/org.0/course_0/Run_0/?page_size=5000', 'num_pages': 2, 'next': None} 2018-07-16 21:52:23.926745+00:00
End collection 2018-07-16 21:52:23.927113+00:00
Query run time: 450.756128s
[OrderedDict([('course_key', u'org.0/course_0/Run_0'), ('completion', OrderedDict([('earned', 522.0), ('possible', 1044.0), ('percent', 0.5)])), ('username', u'robot1')]), OrderedDict([('course_key', u'org.0/course_0/Run_0'), ('completion', OrderedDict([('earned', 522.0), ('possible', 1044.0), ('percent', 0.5)])), ('username', u'robot2')]), OrderedDict([('course_key', u'org.0/course_0/Run_0'), ('completion', OrderedDict([('earned', 522.0), ('possible', 1044.0), ('percent', 0.5)])), ('username', u'robot3')]), OrderedDict([('course_key', u'org.0/course_0/Run_0'), ('completion', OrderedDict([('earned', 522.0), ('possible', 1044.0), ('percent', 0.5)])), ('username', u'robot4')]), OrderedDict([('course_key', u'org.0/course_0/Run_0'), ('completion', OrderedDict([('earned', 522.0), ('possible', 1044.0), ('percent', 0.5)])), ('username', u'robot5')]), OrderedDict([('course_key', u'org.0/course_0/Run_0'), ('completion', OrderedDict([('earned', 522.0), ('possible', 1044.0), ('percent', 0.5)])), ('username', u'robot6')]), OrderedDict([('course_key', u'org.0/course_0/Run_0'), ('completion', OrderedDict([('earned', 522.0), ('possible', 1044.0), ('percent', 0.5)])), ('username', u'robot7')]), OrderedDict([('course_key', u'org.0/course_0/Run_0'), ('completion', OrderedDict([('earned', 522.0), ('possible', 1044.0), ('percent', 0.5)])), ('username', u'robot8')]), OrderedDict([('course_key', u'org.0/course_0/Run_0'), ('completion', OrderedDict([('earned', 522.0), ('possible', 1044.0), ('percent', 0.5)])), ('username', u'robot9')]), OrderedDict([('course_key', u'org.0/course_0/Run_0'), ('completion', OrderedDict([('earned', 522.0), ('possible', 1044.0), ('percent', 0.5)])), ('username', u'robot10')])]
```

**Comment by bradenmacdonald** [comment]
@jcdyer Thanks! Can we also add a similar test including some chapter-level aggregation data, or add that to this test? That's where the most performance issues currently lie, I believe.

**Comment by bradenmacdonald** [comment]
@jcdyer  FYI the docstring at https://github.com/open-craft/openedx-completion-aggregator/blob/3c00fade687aba1f5983c8f411753a20a05c98ed/completion_aggregator/api/v1/views.py#L197-L198 is wrong, as it says that API can only return data for a single user at a time.

**Comment by bradenmacdonald** [comment]
>  If omitted, and the requesting user has staff access, then data for all enrolled users is returned. If the requesting user does not have staff access, then only data for the requesting user is returned.

BTW That API should be changed I think - the information being requested shouldn't vary based on a user's staff status; so if the username is omitted and the user doesn't have staff status, it should raise an error like the course blocks API does.

**Comment by bradenmacdonald** [comment]
@jcdyer 

> It seems to force us to iterate over the queryset for each enrollment, rather than associating each Aggregator with the appropriate enrollment on a single pass over the queryset.

That sounds like an issue, yep. In general for this use case, there won't be many aggregators that don't match active enrollments, so it seems like it would be best to just iterate once over the aggregation results and confirm that the user ID appears in the set of user IDs that are enrolled?

**Comment by jcdyer** [comment]
>That sounds like an issue, yep. In general for this use case, there won't be many aggregators that don't match active enrollments, so it seems like it would be best to just iterate once over the aggregation results and confirm that the user ID appears in the set of user IDs that are enrolled?

That's what I'm thinking.  If the current version improves performance, I don't think we should delay release for it, but I'll write up a ticket for post-release improvements.

**Comment by jcdyer** [comment]
>BTW That API should be changed I think - the information being requested shouldn't vary based on a user's staff status; so if the username is omitted and the user doesn't have staff status, it should raise an error like the course blocks API does.

Agreed.

**Comment by jcdyer** [comment]
@bradenmacdonald It looks like chapter-level aggregation is slowing it down quite a bit.  With 8 chapters in a course, it seems to be on the order of 8 times slower.  I'm checking to make sure having the records in the table doesn't slow down queries that don't request those records.

**Comment by bradenmacdonald** [comment]
@jcdyer Ok, well that would undo the benefits of much of this work, so we need to find an optimization ASAP.

**Comment by jcdyer** [comment]
I'll work on it now.

**Comment by symbolist** [comment]
@jcdyer One nit: The current file path is weird and not idiomatic. Maybe move to `tests/performance/test_views.py`?

**Comment by jcdyer** [comment]
Fixed the weird path.
